scheduler/src/tbf/core.clj

```diff
@@ -0,0 +1,93 @@
+      tokens to be negative. Thus, you can (in one request) spend an unbounded number of tokens. However, you will be
+      in debt until your token balance is at least 0. This design allows the tokens in a request to exceed the max-tokens
+      in the tbf. (If we didn't to it this way, max-tokens is an implicit maximum limit on request size.) By doing it this way
+      we can independently pick the burst size in the tbf and the maximum request size.
+
+      This library does not handle blocking or other operations. Typical workflows are to:
+
+      ** Admission control **
+      This is good when you don't know how big a request will be until afterwards.
+
+      Use in-debt as part of request admission control, rejecting a request if it returns true. When a request completes (or is aborted),
+      take the tokens that it used, affecting the next request.
+
+      ** Delaying requests **
+      If you want to delay requests instead of rejecting. Take the tokens the request needs, pause for the returned pause time.
+      (E.g., schedule the message to be sent at that particular pause time.
+      "
+      ([^Float token-rate ^Long max-tokens ^Long current-time ^Long current-tokens]
+       (assert (> token-rate 0) "Rate must be greater than zero")
+       (assert (>= max-tokens 0) "Max tokens must be greater than zero")
+       (assert (>= current-time 0) "Current time cannot be negative")
+       {:token-rate token-rate :max-tokens max-tokens :last-update current-time :current-tokens current-tokens})
+      ([^Float token-rate ^Long max-tokens ^Long current-time]
+       (make token-rate max-tokens current-time 0)))
+
+(defn- add-tokens-helper
+  "Add tokens (if there are any). If less than a whole token added, keep the current state. Returns the current state.
+  Enforces max-tokens limit. Allows tokens to be removed or added. If no timestamp supplied, just alters the token count."
+  ([{:keys [current-tokens max-tokens] :as tbf} ^Long tokens-to-dispense current-time]
+   (if (= tokens-to-dispense 0)
+     ; Only dispense when we have at least a whole token to add on.
+     tbf
+     (merge tbf
+            {:last-update current-time
+             :current-tokens (min max-tokens (+ current-tokens tokens-to-dispense))})))
+  ([tbf ^Long tokens-to-dispense]
+   (add-tokens-helper tbf tokens-to-dispense (:last-update tbf))))
+
+(defn earn-tokens
+  "Earn tokens. Helper function, takes the current timestamp and adds in any newly earned tokens. It is numerically
+  robust to the timestamp being out of order."
+  [{:keys [last-update token-rate] :as tbf} ^Long current-time]
+  (assert (>= current-time 0))
+  ;(println (str "Foo: " last-update " x " token-rate " y " current-time " TBF: " tbf))
+  (let [current-time (max current-time last-update)]
+    (if (= current-time last-update)
+      tbf
+      (add-tokens-helper tbf (long (* token-rate (- current-time last-update))) current-time))))
+
+(defn time-until
```

Comment:
How do we know what the units are: milliseconds, seconds, minutes?

Reply:
This code doesn't care about the units for measuring time, it can support any of the above.